Updates for Prometheus Exporter

iurygregory opened the pull request:
- Split requirements in test-requirements and requirements
- Update .gitignore
- Initial Parser for IPMI metrics that was collected and tests
(This will be added in another PR: Cable, Drive, Processor, Battery,
 Module, Entity, Add-in, Critical, Voltage, OS, Event, Physical)
- Update Prometheus Driver to generate files with the metrics in
prometheus format
- Create simple driver to be used when we want to collect sample of
data

iurygregory committed:
Updates for Prometheus Exporter

- Split requirements in test-requirements and requirements
- Update .gitignore
- Initial Parser for IPMI metrics that was collected and tests
(This will be added in another PR: Cable, Drive, Processor, Battery,
 Module, Entity, Add-in, Critical, Voltage, OS, Event, Physical)
- Update Prometheus Driver to generate files with the metrics in
prometheus format
- Create simple driver to be used when we want to collect sample of
data

dhellmann commented on ironic_prometheus_exporter/tests/test_driver.py:
```diff
@@ -8,12 +11,65 @@ class TestPrometheusFileNotifier(test_utils.BaseTestCase):
+                     if os.path.isfile(os.path.join(DIR, name))]
+        self.assertEqual(node1, node2)
+        self.assertEqual(len(all_files), 1)
+        self.assertTrue(node1 and node2 in all_files)
```
I don't think this does what you expect. The "node1 and node2" resolves to node2, so it's testing that node2 is in all_files.

```
>>> all_files='ab'
>>> 'a' and 'b' in all_files
True
>>> 'a' and 'b'
'b'
>>>
```

It would be better to compare the all_files list against the expected list of names, in both test cases, to be explicit about the desired content.

iurygregory replied:
Gotcha, going to update

dhellmann commented on ironic_prometheus_exporter/tests/test_driver.py:
```diff
@@ -8,12 +11,65 @@ class TestPrometheusFileNotifier(test_utils.BaseTestCase):
+
+    def test_messages_from_same_node(self):
+        self.config(files_dir='/tmp/ironic_prometheus_exporter',
+                    group='oslo_messaging_notifications')
```
It would be better to use a temporary directory fixture, because that will ensure the data is cleaned up even if the test fails.

iurygregory replied:
Tks Doug

dtantsur replied:
+1

iurygregory replied:
Does the tearDown solve the case ?

dhellmann replied:
Not really. You should use the TempDir fixture from https://pypi.org/project/fixtures/ instead of using a fixed location.

iurygregory replied:
Ok, going to look how to use it =)

iurygregory replied:
I've changed each test for the driver to to have the TempDir and used the value in the driver creation (also I've removed the tearDown method) 
```python
import fixtures

class TestPrometheusFileNotifier(test_utils.BaseTestCase):
    def test_instanciate(self):
        temp_dir = self.useFixture(fixtures.TempDir()).path
        self.config(location=temp_dir,
                    group='oslo_messaging_notifications')

```

iurygregory committed:
Refactoring code

- Extract common functions between IPMI classes
- Update Prometheus driver to not use os.path.dirname in the files_dir
- Update driver test to use tearDown and remove the created files
- Remove unnecessary code in ipmi tests

dtantsur commented on ironic_prometheus_exporter/messaging.py:
```diff
@@ -1,12 +1,13 @@
 prometheus_opts = [
-    cfg.StrOpt('file_path', required=True,
-               help='Path for the json file where the metrics will be stored.')
+    cfg.StrOpt('files_dir', required=True,
```
tiny nit: I'd just call it "location"

iurygregory replied:
Ack

dtantsur commented on ironic_prometheus_exporter/messaging.py:
```diff
@@ -18,13 +19,32 @@ class PrometheusFileDriver(notifier.Driver):
-            raise Exception('The file should end with .json')
-        if not os.path.exists(os.path.dirname(self.file_path)):
-            os.makedirs(os.path.dirname(self.file_path))
+        self.files_dir = conf.oslo_messaging_notifications.files_dir
```
nit: caching the option value like that will prevent it from becoming updateable on SIGHUP. Not sure if you want it, of course..

iurygregory replied:
by 'prevent it from becoming updateable' you mean if someone change the configuration and restart the service we wont have the new value?

dtantsur replied:
I mean, without restarting, with just sending SIGHUP

iurygregory replied:
Not sure the correct way to do this, ideas are welcome =)

dhellmann replied:
In the environment where this is used we don't need to worry about SIGHUP for restarts.

dtantsur commented on ironic_prometheus_exporter/messaging.py:
```diff
@@ -18,13 +19,32 @@ class PrometheusFileDriver(notifier.Driver):
+            node_file.write(node_metrics)
+            node_file.close()
+        except Exception as e:
+            print(e)
```
LOG.error or LOG.exception please. Remember that this will run in the context of ironic-conductor.

iurygregory replied:
Will do

dtantsur commented on ironic_prometheus_exporter/messaging.py:
```diff
@@ -18,13 +19,32 @@ class PrometheusFileDriver(notifier.Driver):
+            node_name = message['payload']['node_name']
+            node_file = open(os.path.join(self.files_dir, node_name), 'w')
+            node_file.write(node_metrics)
+            node_file.close()
```
more idiomatic:
```python
with open(os.path.join(self.files_dir, node_name), 'w') as node_file:
    node_file.write(node_metrics)
```
As a bonus, the file will be closed if write fails.

iurygregory replied:
Will update

dtantsur commented on ironic_prometheus_exporter/parsers/manager.py:
```diff
@@ -0,0 +1,26 @@
+from ironic_prometheus_exporter.parsers import ipmi
+
+
+class ParserManager(object):
```
This class should be a function, unless you plan on having some state for it.

iurygregory replied:
I was thinking as a manager since in the future we will need a parser for Redfish metrics, so based on some information the Manager will choose the correct parser

dtantsur replied:
This sounds like some entry points, but how does a class help? What kind of state will the class maintain?

iurygregory replied:
After we have Redfish data we would be able to load the correct parser here, I didn't understand the state part.

dtantsur commented on ironic_prometheus_exporter/parsers/manager.py:
```diff
@@ -0,0 +1,26 @@
+            ipmi.Memory(payload['Memory'], node_name),
+            ipmi.Power(payload['Power'], node_name),
+            ipmi.Watchdog2(payload['Watchdog2'], node_name),
+            ipmi.Fan(payload['Fan'], node_name)
```
Actually, all these classes should be just functions that you call sequentially.

Unless you want to go fancy and define setuptools entrypoints for each payload item. Even in this case they could be just functions.

iurygregory replied:
you mean change the classes in ipmi module to be functions  and we just call them?

elfosardo replied:
there are very few differences between the classes, I think @dtantsur is right and for example we can just have a list of sensors and call the different functions based on that with little tweaks based on the sensor type, for example if parsing requirements are different

iurygregory replied:
@dtantsur @elfosardo I did some changes let me know what you think =)

iurygregory replied:
Are we good with the changes?

dtantsur commented on test-requirements.txt:
```diff
@@ -0,0 +1,3 @@
+flake8
+stestr>=2.0.0 # Apache-2.0
+oslotest>=3.2.0 # Apache-2.0
```
Why do you need oslotest? For a fancy base unit test class you can use `testtools`. However, you seem to be using the regular `unittest` base class right now.

iurygregory replied:
the driver test requires oslotest to work, otherwise I get ImportError while running the tests

dtantsur replied:
hmm, maybe it's pulled by oslo.messaging. okay.

dtantsur commented on ironic_prometheus_exporter/messaging.py:
```diff
@@ -18,13 +19,32 @@ class PrometheusFileDriver(notifier.Driver):
+    def notify(self, ctx, message, priority, retry):
+        file = open(os.path.join(self.files_dir, 'simplefile'), 'w')
+        file.write(message)
+        file.close()
```
ditto re `with`

iurygregory replied:
ack

dtantsur commented on ironic_prometheus_exporter/messaging.py:
```diff
@@ -18,13 +19,32 @@ class PrometheusFileDriver(notifier.Driver):
+            print(e)
+
+
+class SimpleFileDriver(notifier.Driver):
```
Why do we need this notifier?

iurygregory replied:
I added it so we can collect metrics samples from any source (Redfish or others) since we don't know how they will look like

dtantsur replied:
Can we have a notifier per node? I doubt it. Then you need one notifier to be able to handle all possible drivers more or less gracefully.

iurygregory replied:
I know, the Simple driver is just used to collect the data, after have the data we will update the Prometheus Driver it will call the manager that would handle the correct parser for the metrics and so on.

elfosardo commented on ironic_prometheus_exporter/parsers/ipmi.py:
```diff
@@ -0,0 +1,306 @@
+        self.payload = payload
+        self.node_name = node_name
+
+    def prometheus_format(self):
```
maybe worth rewrite this as common function with parameters?

iurygregory replied:
Will do =)

dhellmann commented on ironic_prometheus_exporter/parsers/ipmi.py:
```diff
@@ -0,0 +1,306 @@
+    metric_dic = {}
+    for entry in payload:
+        if special_label == 'fan':
+            e = re.sub(r'[\d].*$', '', entry.lower())
```
Maybe it's the github PR UI, but I don't see any tests that demonstrate the expected output for all of the parsing in this function. Am I missing it?

iurygregory replied:
You're not missing, I'm going to add some tests for it.

dhellmann commented on ironic_prometheus_exporter/parsers/ipmi.py:
```diff
@@ -0,0 +1,306 @@
+                            'sensor="%s"' % (sensor[0] + sensor_id)]
+            entries_labels[entry] = '{%s}' % ','.join(metric_label)
+        except Exception as e:
+            print(e)
```
I think there was already a comment from @dtantsur about logging, but I don't remember if it was this line or another.

iurygregory replied:
Yeah, going to update =)

dhellmann commented on ironic_prometheus_exporter/parsers/ipmi.py:
```diff
@@ -0,0 +1,306 @@
+
+def extract_labels(entries, payload, node_name):
+    default_label = 'node_name="%s"' % node_name
+    if len(entries) == 1:
```
Why does a len of 1 mean there is no sensor label?

iurygregory replied:
by sensor label you mean "Sensor ID" ?  (Each 
I only added the Sensor ID information as label when the metric name matches for more than 1 Sensor ID (eg. in Temperature we have "Temp (0x1)" and "Temp (0x2)" since the metric name for them is the same the Sensor ID information is added as label.
```
baremetal_temperature_celcius{node_name=...,sensor=Temp1} value1
baremetal_temperature_celcius{node_name=...,sensor=Temp2} value2
```
The len 1 means that the metric only comes from one sensor so I  only need to add the node_name as label. for example:
```
baremetal_exhaust_temperature_celcius{node_name=...} value
```

dhellmann replied:
Ah, ok, that makes sense. Some comments about this sort of thing (and sample data) would make it easier to understand the logic throughout.

iurygregory replied:
I will add

iurygregory committed:
Updates for Driver, Parsers and Tests

-Added logging
-Changed `files_dir` option to `location` for Notification drivers
-Changed write to be done with block
-Extract all classes in ipmi to be functions
-Change manager to extract information based on the event type
-Collect ipmi information only for categories in the payload that
have a match function.
-Update tests for driver to use fixtures TempDir
-Update ipmi parser tests (all categories in one class, test the metric
name for each category)

iurygregory committed:
Update requirements and documentation

- Allow oslo.messaging version to be equal or higher than 9.4.0
- Update documentaion about how to verify if the driver is installed
and how to configure.

dtantsur reviewed (changes requested):
PR #3 includes these changes, which one is correct? Please see my comments there.

iurygregory commented:
@dtantsur probably because I used the the dev branch in my repository as base for the flask_app branch for PR #3

simonpasquier commented on ironic_prometheus_exporter/parsers/ipmi.py:
```diff
@@ -0,0 +1,169 @@
+
+
+def temperature(payload, node_name):
+    available_metrics = metric_names(payload, 'baremetal_', '_celcius')
```
s/celcius/celsius/ ?

simonpasquier commented on ironic_prometheus_exporter/parsers/ipmi.py:
```diff
@@ -0,0 +1,169 @@
+                                use_ipmi_format=use_ipmi_format)
+        if all(v is None for v in values.values()):
+            continue
+        prometheus_info.append(add_prometheus_type(metric, 'gauge'))
```
You would need to add a `#HELP <metric name> description` line to be fully compliant.

simonpasquier commented on ironic_prometheus_exporter/parsers/ipmi.py:
```diff
@@ -0,0 +1,169 @@
+
+def current(payload, node_name):
+    available_metrics = metric_names(payload, 'baremetal_', '')
+    return prometheus_format(payload, node_name, available_metrics,
```
IIUC the metric name doesn't have a [unit](https://prometheus.io/docs/practices/naming/#base-units).

simonpasquier commented on ironic_prometheus_exporter/parsers/ipmi.py:
```diff
@@ -0,0 +1,169 @@
+    return metric_dic
+
+
+def extract_labels(entries, payload, node_name):
```
I'm wondering whether having a label storing `Entity ID` would be worthwile too?

simonpasquier commented on ironic_prometheus_exporter/parsers/ipmi.py:
```diff
@@ -0,0 +1,169 @@
+            sensor = payload[entry]['Sensor ID'].split()
+            sensor_id = str(int(re.sub(r'[\(\)]', '', sensor[-1]), 0))
+            metric_label = [default_label,
+                            'sensor="%s"' % (sensor[0] + sensor_id)]
```
Why not storing the `Sensor ID` value as-is? Label values can contain any valid UTF-8 character.

simonpasquier commented on ironic_prometheus_exporter/parsers/ipmi.py:
```diff
@@ -0,0 +1,169 @@
+            sensor_id = str(int(re.sub(r'[\(\)]', '', sensor[-1]), 0))
+            metric_label = [default_label,
+                            'sensor="%s"' % (sensor[0] + sensor_id)]
+            entries_labels[entry] = '{%s}' % ','.join(metric_label)
```
For safety, you might want to replace any character not matching `[a-zA-Z0-9_]` by `_` and check that the first character isn't a digit (see https://prometheus.io/docs/concepts/data_model/).

simonpasquier commented on ironic_prometheus_exporter/parsers/ipmi.py:
```diff
@@ -0,0 +1,169 @@
+            if 'mem' not in label and 'memory' not in label:
+                label = 'memory_' + label
+
+        metric_name = prefix + label + sufix
```
For safety, you might want to replace any character not matching `[a-zA-Z0-9_:]` by `_` and check that the first character isn't a digit (see https://prometheus.io/docs/concepts/data_model/).

simonpasquier reviewed (commented):
Have you evaluated using the official [client_python](https://github.com/prometheus/client_python) library? This would help catching erroneous metrics and ensure that the generated data is fully compliant with the Prometheus exposition format. FWIW it also provides a [helper](https://github.com/prometheus/client_python/#node-exporter-textfile-collector) to dump metrics to text files.

iurygregory commented:
Hi @simonpasquier thanks for the review I will take a look in the weekend since I'm travelling today, can you check the PR #3 it contains the commits from this PR and the flask application. If you spot other things that would be worth to change. I will look at the python library. Thanks

iurygregory commented:
The requests are addressed on PR #3 I'm going to close this one
